jetson/percep_obs_detect/src/ar_tag/artag_detector.cpp

```diff
@@ -0,0 +1,215 @@
+        if (t0.loc.x < t1.loc.x) {  //if tag 0 is left of tag 1, put t0 first
+            discoveredTags.first = t0;
+            discoveredTags.second = t1;
+        } else {  //tag 1 is left of tag 0, put t1 first
+            discoveredTags.first = t1;
+            discoveredTags.second = t0;
+        }
+    } else {  // detected >=3 tags
+        // return leftmost and rightsmost detected tags to account for potentially seeing 2 of each tag on a post
+        Tag t0, t1;
+        t0.id = ids[0];
+        t0.loc = getAverageTagCoordinateFromCorners(corners[0]);
+        t1.id = ids[ids.size() - 1];
+        t1.loc = getAverageTagCoordinateFromCorners(corners[ids.size() - 1]);
+        if (t0.loc.x < t1.loc.x) {  //if tag 0 is left of tag 1, put t0 first
+            discoveredTags.first = t0;
+            discoveredTags.second = t1;
+        } else {  //tag 1 is left of tag 0, put t1 first
+            discoveredTags.first = t1;
+            discoveredTags.second = t0;
+        }
+    }
+    return discoveredTags;
+}
+
+double TagDetector::getAngle(float xPixel, float wPixel){
+    double fieldofView = 110 * PI/180;
+    return atan((xPixel - wPixel/2)/(wPixel/2)* tan(fieldofView/2))* 180.0 /PI;
+}
+
+void TagDetector::updateDetectedTagInfo(rover_msgs::Target *arTags, pair<Tag, Tag> &tagPair, Mat &depth_img, Mat &src){
+    struct tagPairs {
+        vector<int> id;
+        vector<int> locx;
+        vector<int> locy;
+        vector<int> buffer;
+    }; 
+    tagPairs tags;
+
+    tags.id.push_back(tagPair.first.id);
+    tags.locx.push_back(tagPair.first.loc.x);
+    tags.locy.push_back(tagPair.first.loc.y);
+    tags.id.push_back(tagPair.second.id);
+    tags.locx.push_back(tagPair.second.loc.x);
+    tags.locy.push_back(tagPair.second.loc.y);
+    tags.buffer.push_back(0);
+    tags.buffer.push_back(0);
+
+  for (uint i=0; i<2; i++) {
+    if(tags.id[i] == DEFAULT_TAG_VAL){//no tag found
+        if(tags.buffer[i] <= BUFFER_ITERATIONS){//send buffered tag until tag is found
+            ++tags.buffer[i];
+        } else {//if still no tag found, set all stats to -1
+            arTags[i].distance = DEFAULT_TAG_VAL;
+            arTags[i].bearing = DEFAULT_TAG_VAL;
+            arTags[i].id = DEFAULT_TAG_VAL;
+        }
+     } else {//tag found
+    if(!isnan(depth_img.at<float>(tags.locy.at(i), tags.locx.at(i)))) {
+        arTags[i].distance = depth_img.at<float>(tags.locy.at(i), tags.locx.at(i)) / MM_PER_M;
+    }
+        arTags[i].bearing = getAngle((int)tags.locx.at(i), src.cols);
+        arTags[i].id = tags.id.at(i);
+        tags.buffer[i] = 0;
+   }
+  }
+}
+
+void TagDetector::update() {
+    // Initializations
+    #ifndef NO_JARVIS
+    lcm::LCM lcm_;
+    #endif
+    rover_msgs::TargetList arTagsMessage;
+    rover_msgs::Target* arTags = arTagsMessage.targetList;
+    arTags[0].distance = DEFAULT_TAG_VAL;
+    arTags[1].distance = DEFAULT_TAG_VAL;
+    
+    Mat rgb;
+    Mat src = cam->get_image();
```

Comment:
I'm not exactly sure how semantics work for the cv Mat type, but could this be reference?

Reply:
Changed, but untested